File: BST/InorderSuccPred/sol.py
class Solution:
    def prepareInVector(self, root, ans):
        if not root:
            return

        self.prepareInVector(root.left, ans)
        ans.append(root)
        self.prepareInVector(root.right, ans)

    def findPreSuc(self, root, key):
        check = []
        self.prepareInVector(root, check)

        pred = None
        succ = None

        for node in check:
            val = node.data

            if key > val:
                pred = node
            elif key < val:
                succ = node
                break

        return [pred, succ]

File: BST/InorderSuccPred/sol.cpp
#include <iostream>
#include <vector>
using namespace std;

class Node {
   public:
    int data;
    Node *left;
    Node *right;

    Node(int x){
        data = x;
        left = NULL;
        right = NULL;
    }
};

class Solution {
  public:
    void prepareInVector(Node* root,vector<Node*>& ans){
        if(!root)return;
        prepareInVector(root->left,ans);
        ans.push_back(root);
        prepareInVector(root->right,ans);
    }
    vector<Node*> findPreSuc(Node* root, int key) {
        // code here
        vector<Node*>check;
        prepareInVector(root,check);
        int n = check.size();
        Node* pred = NULL;
        Node* succ = NULL;
        for(int i=0;i<n;i++){
            int val = check[i]->data;
            if(key > val){
                pred = check[i];
            }else{
                if(key < val){
                    succ = check[i];
                    break;
                }
            }
        }
        return {pred,succ};
    }
};